File: task_management_app/lib/models/task_model.dart
class Task {
  final String urn;
  final String taskName;
  final String assignedBy;
  final String assignedTo;
  final DateTime commencementDate;
  final DateTime dueDate;
  final String clientName;
  final bool status;
  final String taskStatus;

  Task({
    required this.urn,
    required this.taskName,
    required this.assignedBy,
    required this.assignedTo,
    required this.commencementDate,
    required this.dueDate,
    required this.clientName,
    this.status = true,
    this.taskStatus = "Scheduled",
  });

  Map<String, dynamic> toJson() {
    return {
      // "urn": urn,
      "taskName": taskName,
      "assignedBy": assignedBy,
      "assignedTo": assignedTo,
      "commencementDate": commencementDate.toIso8601String(),
      "dueDate": dueDate.toIso8601String(),
      "clientName": clientName,
      "status": status,
      "taskStatus": taskStatus,
    };
  }
}

File: server/models/taskModel.js
const mongoose = require('mongoose');

const TaskSchema = new mongoose.Schema(
    {
        urn: { type: String, unique: true, index: true }, 
        taskName: { type: String, required: true, trim: true },
        assignedBy: { type: String, required: true, trim: true },
        assignedTo: { type: String, required: true, trim: true },
        commencementDate: { type: Date, required: true },
        dueDate: { type: Date, required: true },
        clientName: { type: String, required: true, trim: true },
        status: { type: Boolean, default: true }, 
        taskStatus: {
            type: String,
            required: true,
            enum: ["Scheduled", "In Progress", "Completed", "With-held"], 
            default: "Scheduled"
        }
    },
    { timestamps: true } 
);


TaskSchema.pre("save", async function (next) {
    if (!this.urn) {
        const lastTask = await mongoose.model("tasks").findOne().sort({ createdAt: -1 });
        const lastUrn = lastTask ? parseInt(lastTask.urn.replace("TASK", ""), 10) : 0;
        this.urn = `TASK${(lastUrn + 1).toString().padStart(3, '0')}`;
    }
    next();
});

module.exports = mongoose.model("tasks", TaskSchema);
